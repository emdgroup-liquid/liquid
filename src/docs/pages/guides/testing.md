---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

