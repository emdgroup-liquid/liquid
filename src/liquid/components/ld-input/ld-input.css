/*
:where(.ld-input) {
  width: 12rem;
}
*/

.ld-input {
  --ld-input-padding-x: 0.625rem;
  --ld-input-padding-top: 0.625rem;
  --ld-input-padding-bottom: 0.6875rem;
  --ld-input-max-height: 2.5rem;
  display: inline-flex;
  align-items: center;
  background-color: var(--ld-col-bg-g);
  color: var(--ld-col-rblck-default);
  box-shadow: inset 0 0 0 0.1rem var(--ld-col-rblck1);
  max-width: 100%;
  border-radius: var(--ld-sp-4);
  line-height: 1;

  > :where(:not(input)) {
    user-select: none;
  }

  > :where(:not(input):first-child) {
    margin-left: var(--ld-sp-12);
  }

  > :where(:not(input):last-child) {
    margin-right: var(--ld-sp-12);
  }

  > input {
    padding: var(--ld-input-padding-top) var(--ld-input-padding-x)
      var(--ld-input-padding-bottom);
    font: var(--ld-typo-body-m);
    line-height: 1;
    background-color: transparent;
    width: 100%;
    border: 0;
    border-radius: var(--ld-sp-4);
    max-height: var(--ld-input-max-height);
    -webkit-appearance: none;

    &::placeholder {
      color: var(--ld-col-rblck4);
    }

    &[type='search'] {
      &::-webkit-search-decoration,
      &::-webkit-search-cancel-button,
      &::-webkit-search-results-button,
      &::-webkit-search-results-decoration {
        -webkit-appearance: none;
      }
    }
  }

  > ld-button,
  > .ld-button {
    background-color: transparent;

    &:first-child,
    &:first-child > button {
      margin-left: 0;

      &:not(.ld-button--sm) {
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
      }

      &.ld-button--sm {
        margin: var(--ld-sp-4);
        margin-right: 0;
      }
    }

    &:last-child,
    &:last-child > button {
      margin-right: 0;

      &:not(.ld-button--sm) {
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
      }

      &.ld-button--sm {
        margin: var(--ld-sp-4);
        margin-left: 0;
      }
    }

    &[mode='ghost'],
    &.ld-button--ghost {
      &:first-child {
        margin-right: calc(-1 * var(--ld-input-padding-x));
      }

      &:last-child {
        margin-left: calc(-1 * var(--ld-input-padding-x));
      }
    }
  }
  > ld-button.hydrated > button.ld-button--ghost,
  > button.ld-button.ld-button--ghost {
    @media (hover: hover) {
      &:hover {
        background-color: transparent;
      }
    }
    &:active,
    &:active:focus-visible,
    &:focus:focus-visible {
      background-color: transparent;
    }
  }

  > ld-icon,
  > .ld-icon {
    color: var(--ld-col-rb-default);
    display: inline-flex;
  }

  .ld-theme-solvent &,
  .ld-theme-bubblegum &,
  .ld-theme-shake & {
    > ld-icon,
    > .ld-icon {
      color: var(--ld-col-rp-default);
    }
  }

  .ld-theme-tea & {
    > ld-icon,
    > .ld-icon {
      color: var(--ld-col-rg-default);
    }
  }

  .ld-label & {
    width: 100%;
  }
}

.ld-input--light {
  background-color: var(--ld-col-wht);
}

:where(.ld-input:not([disabled]:not([aria-disabled]))) {
  &:hover {
    box-shadow: inset 0 0 0 0.1rem var(--ld-col-rblck2);
  }
}

:where(.ld-input) {
  &[disabled],
  &[aria-disabled] {
    box-shadow: inset 0 0 0 0.1rem var(--ld-col-rblck1);

    > input {
      color: var(--ld-col-rblck2);
      caret-color: transparent;

      &::placeholder {
        opacity: 0.5;
      }
    }
  }
  &:focus-within {
    box-shadow: inset 0 0 0 0.1rem var(--ld-col-rb-default);

    &:not([disabled]):not([aria-disabled]) {
      background-color: var(--ld-col-wht);
    }

    > input {
      outline: none;
    }

    .ld-theme-solvent &,
    .ld-theme-bubblegum &,
    .ld-theme-shake & {
      box-shadow: inset 0 0 0 0.1rem var(--ld-col-rp-default);
    }
    .ld-theme-tea & {
      box-shadow: inset 0 0 0 0.1rem var(--ld-col-rg-default);
    }
  }
}

.ld-input--invalid {
  box-shadow: inset 0 0 0 0.1rem var(--ld-col-rr-default);

  &:where(:not(:focus)) {
    color: var(--ld-col-rr-default);
    background-color: var(--ld-col-rr1);

    &::placeholder {
      color: var(--ld-col-rr3);
    }
  }
}
