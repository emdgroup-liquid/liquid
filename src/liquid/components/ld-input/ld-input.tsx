import {
  Component,
  Element,
  h,
  Host,
  Method,
  Prop,
  State,
  Watch,
} from '@stencil/core'
import { cloneAttributes } from '../../utils/cloneAttributes'
import { getClassNames } from '../../utils/getClassNames'

/**
 * The `ld-input` component. You can use it in conjunction with the `ld-label`
 * and the `ld-input-message` component. See examples in the docs for a better
 * understanding on how they can be used together.
 *
 * @slot start - The purpose of this slot is to add icons or buttons
 * to the input, __justifying the item to the end of the component__.
 * Styling for `ld-icon` and `ld-button` is provided within the `ld-input` component.
 * If you choose to place something different into the slot, you will probably
 * need to adjust some styles on the slotted item in order to make it fit right.
 * @slot end - The purpose of this slot is to add icons or buttons
 * to the input, __justifying the item to the start of the component__.
 * Styling for `ld-icon` and `ld-button` is provided within the `ld-input` component.
 * If you choose to place something different into the slot, you will probably
 * need to adjust some styles on the slotted item in order to make it fit right.
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part input - Actual input/textarea element
 * @part placeholder - Placeholder rendered for input type "file"
 */
@Component({
  tag: 'ld-input',
  styleUrl: 'ld-input.css',
  shadow: true,
})
export class LdInput implements InnerFocusable, ClonesAttributes {
  @Element() el: HTMLInputElement | HTMLTextAreaElement

  private attributesObserver: MutationObserver

  private hiddenInput?: HTMLInputElement
  private input: HTMLInputElement | HTMLTextAreaElement

  /** Hint for expected file type in file upload controls. */
  @Prop() accept?: string

  /** Hint for form autofill feature. */
  @Prop({ mutable: true, reflect: true }) autocomplete?: string

  /** Automatically focus the form control when the page is loaded. */
  @Prop() autofocus?: boolean

  /** Media capture input method in file upload controls. */
  @Prop() capture?: string

  /** The number of columns. */
  @Prop() cols?: number

  /** Name of form field to use for sending the element's directionality in form submission. */
  @Prop() dirname?: string

  /** Whether the form control is disabled. */
  @Prop() disabled?: boolean

  /** Associates the control with a form element. */
  @Prop() form?: string

  /** Set this property to `true` in order to mark the field visually as invalid. */
  @Prop() invalid?: boolean

  /** Value of the id attribute of the `<datalist>` of autocomplete options. */
  @Prop() list?: string

  /** Maximum value. */
  @Prop() max?: string

  /** Maximum length (number of characters) of `value`. */
  @Prop() maxlength?: string

  /** Minimum value. */
  @Prop() min?: string

  /** Minimum length (number of characters) of `value`. */
  @Prop() minlength?: string

  /**
   * Uses textarea instead of input internally. Setting this attribute to true
   * disables the attribute type and both slots.
   */
  @Prop() multiline?: boolean

  /** Boolean. Whether to allow multiple values. */
  @Prop() multiple?: boolean

  /** Used to specify the name of the control. */
  @Prop() name?: string

  /** Pattern the `value` must match to be valid. */
  @Prop() pattern?: string

  /** The input placeholder. */
  @Prop() placeholder?: string

  /** The value is not editable. */
  @Prop() readonly?: boolean

  /** A value is required for the form to be submittable. */
  @Prop() required?: boolean

  /** The number of rows. */
  @Prop() rows?: number

  /** Size of the input. */
  @Prop() size?: 'sm' | 'lg'

  /** Incremental values that are valid. */
  @Prop() step?: string

  /** Input tone. Use `'dark'` on white backgrounds. Default is a light tone. */
  @Prop() tone?: 'dark'

  /** The input type. */
  @Prop() type: string

  /** The input value. */
  @Prop({ mutable: true }) value?: string

  @State() clonedAttributes

  /**
   * Sets focus on the input
   */
  @Method()
  async focusInner() {
    if (this.input !== undefined) {
      this.input.focus()
    }
  }

  @Watch('dirname')
  @Watch('form')
  @Watch('name')
  @Watch('value')
  updateHiddenInput() {
    const outerForm = this.el.closest('form')
    if (!this.hiddenInput && this.name && (outerForm || this.form)) {
      this.createHiddenInput()
    }

    if (this.hiddenInput) {
      if (this.dirname) {
        this.hiddenInput.dirName = this.dirname
      }

      if (this.name) {
        this.hiddenInput.name = this.name
      } else if (this.hiddenInput.name) {
        this.hiddenInput.remove()
        this.hiddenInput = undefined
        return
      }

      if (this.form) {
        this.hiddenInput.setAttribute('form', this.form)
      } else if (this.hiddenInput.getAttribute('form')) {
        if (outerForm) {
          this.hiddenInput.removeAttribute('form')
        } else {
          this.hiddenInput.remove()
          this.hiddenInput = undefined
          return
        }
      }

      if (this.value) {
        this.hiddenInput.value = this.value
      } else if (this.hiddenInput.value) {
        this.hiddenInput.removeAttribute('value')
      }
    }
  }

  private createHiddenInput() {
    this.hiddenInput = document.createElement('input')
    this.hiddenInput.type = 'hidden'
    this.el.appendChild(this.hiddenInput)
  }

  componentWillLoad() {
    this.attributesObserver = cloneAttributes.call(this, [
      'multiline',
      'autocomplete',
    ])

    const outerForm = this.el.closest('form')

    if (outerForm && !this.autocomplete) {
      this.autocomplete = outerForm.getAttribute('autocomplete')
    }

    if (this.name && (outerForm || this.form)) {
      this.createHiddenInput()
      this.hiddenInput.name = this.name

      if (this.dirname) {
        this.hiddenInput.dirName = this.dirname
      }

      if (this.form) {
        this.hiddenInput.setAttribute('form', this.form)
      }

      if (this.value) {
        this.hiddenInput.value = this.value
      }
    }

    this.el.querySelectorAll('ld-button').forEach((button) => {
      if (this.size !== undefined) {
        button.setAttribute('size', this.size)
      } else {
        button.removeAttribute('size')
      }
    })
    this.el.querySelectorAll('.ld-button').forEach((button) => {
      if (this.size === 'sm') {
        button.classList.remove('ld-button--lg')
        button.classList.add('ld-button--sm')
      } else if (this.size === 'lg') {
        button.classList.remove('ld-button--sm')
        button.classList.add('ld-button--lg')
      } else {
        button.classList.remove('ld-button--sm', 'ld-button--lg')
      }
    })
    this.el.querySelectorAll('ld-icon').forEach((icon) => {
      if (this.size !== undefined) {
        icon.setAttribute('size', this.size)
      } else {
        icon.removeAttribute('size')
      }
    })
    this.el.querySelectorAll('.ld-icon').forEach((icon) => {
      if (this.size === 'sm') {
        icon.classList.remove('ld-icon--lg')
        icon.classList.add('ld-icon--sm')
      } else if (this.size === 'lg') {
        icon.classList.remove('ld-icon--sm')
        icon.classList.add('ld-icon--lg')
      } else {
        icon.classList.remove('ld-icon--sm', 'ld-icon--lg')
      }
    })
  }

  componentDidLoad() {
    if (this.autofocus) {
      this.focusInner()
    }
  }

  private handleBlur = (ev: FocusEvent) => {
    setTimeout(() => {
      this.el.dispatchEvent(ev)
    })
  }

  private handleFocus = (ev: FocusEvent) => {
    setTimeout(() => {
      this.el.dispatchEvent(ev)
    })
  }

  private handleInput() {
    if (!(this.input.getAttribute('aria-disabled') === 'true')) {
      this.value = this.input.value
      return
    }
    this.input.value = this.value ?? ''
  }

  private handleClick = (ev: MouseEvent) => {
    const target = ev.target as HTMLElement
    if (
      this.el.hasAttribute('disabled') ||
      this.el.getAttribute('aria-disabled') === 'true'
    ) {
      ev.preventDefault()
      return
    }

    if (target.closest('ld-button')) return

    if (target === this.el) {
      this.input.focus()
      this.input.dispatchEvent(new Event('click', { bubbles: false }))
    } else {
      this.input.focus()
    }
  }

  private handleKeyDown = (ev: KeyboardEvent) => {
    const outerForm = this.el.closest('form')
    const formToSubmit = this.form
      ? document.querySelector<HTMLFormElement>(`#${this.form}`) ?? outerForm
      : outerForm

    if (
      this.el.getAttribute('aria-disabled') === 'true' &&
      !['ArrowLeft', 'ArrowRight', 'Tab'].includes(ev.key)
    ) {
      ev.preventDefault()
    } else if (!this.multiline && ev.key === 'Enter' && formToSubmit) {
      formToSubmit.requestSubmit()
    }
  }

  disconnectedCallback() {
    this.attributesObserver?.disconnect()
  }

  render() {
    const cl = getClassNames([
      'ld-input',
      this.size && `ld-input--${this.size}`,
      this.tone && `ld-input--${this.tone}`,
      this.invalid && 'ld-input--invalid',
    ])

    if (this.multiline) {
      const { type, ...clonedAttributesWithoutType } = this.clonedAttributes
      return (
        <Host class={cl} onClick={this.handleClick}>
          <textarea
            {...clonedAttributesWithoutType}
            onBlur={this.handleBlur}
            onFocus={this.handleFocus}
            onInput={this.handleInput.bind(this)}
            part="input focusable"
            ref={(el) => (this.input = el)}
            value={this.value}
          />
          {type === 'file' && (
            <span class="ld-input__placeholder" part="placeholder">
              {this.input?.value || this.placeholder}
            </span>
          )}
        </Host>
      )
    }

    return (
      <Host class={cl} onClick={this.handleClick}>
        <slot name="start"></slot>
        <input
          {...this.clonedAttributes}
          autocomplete={this.autocomplete}
          onBlur={this.handleBlur}
          onFocus={this.handleFocus}
          onInput={this.handleInput.bind(this)}
          onKeyDown={this.handleKeyDown}
          part="input focusable"
          ref={(el) => (this.input = el)}
          value={this.value}
        />
        {this.type === 'file' && (
          <span class="ld-input__placeholder" part="placeholder">
            {this.input?.value || this.placeholder}
          </span>
        )}
        <slot name="end"></slot>
      </Host>
    )
  }
}
