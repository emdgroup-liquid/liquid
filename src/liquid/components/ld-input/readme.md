---
eleventyNavigation:
  key: Input
  parent: Components
layout: layout.njk
title: Input
permalink: liquid/components/ld-input/
---

# ld-input

---

## Web component

{% example %}
<ld-input placeholder="Placeholder"></ld-input>
<ld-input value="Value"></ld-input>
{% endexample %}

### Disabled

{% example %}
<ld-input placeholder="Placeholder" disabled></ld-input>
<ld-input disabled value="Value"></ld-input>
{% endexample %}

### Light mode

{% example %}
<ld-input mode="light" placeholder="Placeholder"></ld-input>
<ld-input mode="light" value="Value"></ld-input>
{% endexample %}

### Light mode, disabled

{% example %}
<ld-input mode="light" disabled placeholder="Placeholder"></ld-input>
<ld-input mode="light" disabled value="Value"></ld-input>
{% endexample %}

### With label

{% example %}
<ld-label>
  Email Address
  <ld-input placeholder="jane.doe@example.com"></ld-input>
</ld-label>
{% endexample %}

Please reffer to the [ld-label](/liquid/components/ld-label/) docs for more information on the label component.

### With label and input message

{% example %}
<div style="display: grid; gap: 1rem; grid-template-columns: repeat(auto-fit, minmax(12rem, 1fr)); width: 100%;">
  <ld-label>
    Email Address
    <ld-input invalid placeholder="jane.doe@example.com" value="yolo"></ld-input>
    <ld-input-message>The email address is invalid.</ld-input-message>
  </ld-label>

  <ld-label>
    Password
    <ld-input type="password" value="asdf1234" type password></ld-input>
    <ld-input-message mode="info">Use at least one special character (~!@#$%^&*_-+=|\(){}[]:;<>,.?/)</ld-input-message>
  </ld-label>
</div>
{% endexample %}

#### On input message width and placement

By default, the input field stretches to the maximum width of its wrapping label, which in turn stretches to the width of its content (the computed width of the `ld-input-message` component). So, if you want to set certain constraints on the width of the all three components, you can do so by setting the constraint on the wrapping element – the `ld-label` component.

{% example %}
<ld-label>
  Email Address
  <ld-input placeholder="jane.doe@example.com"></ld-input>
  <ld-input-message mode="info">This info message is extremely long and makes all three components (the label, the input and itself) grow horizontaly.</ld-input-message>
</ld-label>

<ld-label style="max-width: 20rem">
  Email Address
  <ld-input placeholder="jane.doe@example.com"></ld-input>
  <ld-input-message mode="info">This info message is also extremely long, but since the label has a max width, all three components (the label, the input and itself) can take only the maximum width of the label.</ld-input-message>
</ld-label>
{% endexample %}

When displaying input messages conditionally (i.e. an error message becomes visible as soon as an input has been interacted with, but the value is still invalid) you should try to position UI elements in a way that prevents [layout shifts](https://web.dev/cls/). For instance, you can “reserve space” for your messages and then make them appear in the reserved space without pushing other content to the bottom (i.e. using `position: absolute` or some “flexy” layout). Needless to say, results look best if you keep the messages short.

### Example vanilla JS form validation

{% example %}
<style>
#example-form {
  display: flex;
  flex-wrap: wrap;
  gap: 1rem;
  width: 100%;
}
#example-form > * {
  align-self: flex-end;
  flex: 1 0 auto;
}
#example-form ld-button {
  margin-bottom: 1.6rem;
}
@media (max-width: 52rem) {
  #example-form > * {
    width: 100%;
  }
}
#example-form ld-input-message {
  visibility: hidden;
}
#example-form ld-input-message[covert="false"] {
  visibility: inherit;
}
</style>
<form id="example-form" novalidate>
  <ld-label>
    Login*
    <ld-input required placeholder="login"></ld-input>
    <ld-input-message visible="false">This field is required.</ld-input-message>
  </ld-label>
  <ld-label>
    Password*
    <ld-input required placeholder="password" type="password"></ld-input>
    <ld-input-message visible="false">This field is required.</ld-input-message>
  </ld-label>
  <ld-button>Submit</ld-button>
</form>
<script>
  const form = document.getElementById('example-form')
  const username = document.querySelector('#example-form ld-label:first-of-type ld-input')
  const usernameErrorMessage = document.querySelector('#example-form ld-label:first-of-type ld-input-message')
  const password = document.querySelector('#example-form ld-label:last-of-type ld-input')
  const passwordErrorMessage = document.querySelector('#example-form ld-label:last-of-type ld-input-message')
  async function validateInput(ldInput, ldInputMessage) {
    const value = await ldInput.getValue()
    if (!value) {
      ldInput.setAttribute('invalid', 'true')
      ldInputMessage.setAttribute('covert', 'false')
    } else {
      ldInput.removeAttribute('invalid')
      ldInputMessage.setAttribute('covert', 'true')
    }
  }
  username.addEventListener('input', ev => {
    validateInput(username, usernameErrorMessage)
  })
  username.addEventListener('blur', ev => {
    validateInput(username, usernameErrorMessage)
  })
  password.addEventListener('blur', ev => {
    validateInput(password, passwordErrorMessage)
  })
  password.addEventListener('input', ev => {
    validateInput(password, passwordErrorMessage)
  })
</script>
{% endexample %}


<!-- Auto Generated Below -->


## Properties

| Property  | Attribute | Description                                                                        | Type                | Default     |
| --------- | --------- | ---------------------------------------------------------------------------------- | ------------------- | ----------- |
| `invalid` | `invalid` | Set this property to `true` in order to mark the field visually as invalid.        | `boolean`           | `undefined` |
| `mode`    | `mode`    | Input mode. Use `'dark'` on white backgrounds, use `'light'` on other backgrounds. | `"dark" \| "light"` | `'dark'`    |


## Methods

### `getValue() => Promise<string>`

Returns input value.

#### Returns

Type: `Promise<string>`




----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
