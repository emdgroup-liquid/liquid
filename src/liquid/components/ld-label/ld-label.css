:host {
  display: inline-flex;

  .ld-label {
    width: 100%;
  }
}

.ld-label {
  /* layout */
  --ld-label-column-gap: 0.625rem;
  --ld-label-row-gap: var(--ld-sp-6);

  /* colors */
  --ld-label-col: var(--ld-col-neutral);
  --ld-label-disabled-col: var(--ld-col-neutral-300);

  color: var(--ld-label-col);
  font: var(--ld-typo-label-s);
  display: inline-grid;
  row-gap: var(--ld-label-row-gap);
  column-gap: 0;
  max-width: 100%;
  align-content: flex-start;
}

.ld-label--left,
.ld-label--right {
  justify-content: flex-start;
  align-items: center;
  column-gap: var(--ld-label-column-gap);
  row-gap: 0;

  > ::slotted(.ld-checkbox),
  > .ld-checkbox,
  > ::slotted(.ld-radio),
  > .ld-radio,
  > ::slotted(.ld-toggle),
  > .ld-toggle,
  > ::slotted(.ld-input),
  > .ld-input,
  > ::slotted(.ld-select),
  > .ld-select {
    grid-area: input;
    align-self: baseline;
  }

  > ::slotted(.ld-input-message),
  > .ld-input-message {
    grid-area: message;
    margin-top: var(--ld-sp-4);
  }
}

.ld-label--align-message {
  &.ld-label--left {
    > ::slotted(.ld-input-message),
    > .ld-input-message {
      grid-column-start: 2;
    }
  }

  &.ld-label--right {
    > ::slotted(.ld-input-message),
    > .ld-input-message {
      grid-column-end: 2;
    }
  }
}

.ld-label--left {
  grid-template-columns: 1fr auto;
  grid-template-areas:
    'label input'
    'message message';
}

.ld-label--right {
  grid-template-columns: auto 1fr;
  grid-template-areas:
    'input label'
    'message message';
}

.ld-label--m {
  font: var(--ld-typo-label-m);
  line-height: 1;
}

:where(.ld-label[disabled]) {
  color: var(--ld-label-disabled-col);
}
