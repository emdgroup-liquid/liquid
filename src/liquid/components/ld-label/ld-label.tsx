import { Component, Element, h, Prop, State } from '@stencil/core'
import { cloneAttributes } from '../../utils/cloneAttributes'
import { getClassNames } from '../../utils/getClassNames'

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part tag - Actual label element
 */
@Component({
  tag: 'ld-label',
  styleUrl: 'ld-label.css',
  shadow: true,
})
export class LdLabel implements ClonesAttributes {
  @Element() el: HTMLLabelElement

  private attributesObserver: MutationObserver

  /** Align input message with input position. */
  @Prop() alignMessage: boolean

  /** Relative position to labeled element. Default is top. */
  @Prop() position: 'left' | 'right'

  /** Size of the label. Default is small. */
  @Prop() size: 'm'

  @State() clonedAttributes

  private handleClick = async (event: MouseEvent) => {
    const inputElement: HTMLElement = this.el.querySelector(
      'ld-input, ld-textarea, ld-toggle, ld-select, ld-button, ld-checkbox, ld-radio, input, textarea, button, select'
    )
    const clickedInsideInputElement =
      event.target === inputElement ||
      inputElement.contains(event.target as Node)

    if (
      inputElement &&
      !clickedInsideInputElement &&
      !inputElement['disabled']
    ) {
      if ('focusInner' in inputElement) {
        await (inputElement as unknown as InnerFocusable).focusInner()
      } else {
        inputElement.focus()
      }

      inputElement.click()
    }
  }

  componentWillLoad() {
    this.attributesObserver = cloneAttributes.call(this, [
      'align-message',
      'position',
      'size',
    ])
  }

  disconnectedCallback() {
    this.attributesObserver?.disconnect()
  }

  render() {
    const cl = getClassNames([
      'ld-label',
      this.alignMessage && 'ld-label--align-message',
      this.position && `ld-label--${this.position}`,
      this.size && `ld-label--${this.size}`,
    ])

    return (
      <label
        {...this.clonedAttributes}
        class={cl}
        onClick={this.handleClick}
        part="tag"
      >
        <slot></slot>
      </label>
    )
  }
}
