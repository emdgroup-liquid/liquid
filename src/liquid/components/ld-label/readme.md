---
eleventyNavigation:
  key: Label
  parent: Components
layout: layout.njk
title: Label
permalink: liquid/components/ld-label/
---

# ld-label

This component is meant to be used in conjunction with the [`ld-input`](/liquid/components/ld-input/) and [`ld-input-message`](/liquid/components/ld-input-message/) component. Please reffer to the [`ld-input`](/liquid/components/ld-input/) docs for further usage examples.

## Example

{% example %}
<ld-label>Email Address</ld-label>

<!-- CSS component -->

<label class="ld-label">Email Address</label>
{% endexample %}


<!-- Auto Generated Below -->


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
