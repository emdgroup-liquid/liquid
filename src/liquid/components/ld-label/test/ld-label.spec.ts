import { newSpecPage } from '@stencil/core/testing'
import { LdInput } from '../../ld-input/ld-input'
import { LdLabel } from '../ld-label'

describe('ld-label', () => {
  it('renders with slot', async () => {
    const { root } = await newSpecPage({
      components: [LdLabel],
      html: `<ld-label>Yada-yada</ld-label>`,
    })
    expect(root).toMatchSnapshot()
  })

  it('renders with position left', async () => {
    const { root } = await newSpecPage({
      components: [LdLabel],
      html: `<ld-label position="left">Yada-yada</ld-label>`,
    })
    expect(root).toMatchSnapshot()
  })

  it('renders with position right', async () => {
    const { root } = await newSpecPage({
      components: [LdLabel],
      html: `<ld-label position="right">Yada-yada</ld-label>`,
    })
    expect(root).toMatchSnapshot()
  })

  it('renders with size "m"', async () => {
    const { root } = await newSpecPage({
      components: [LdLabel],
      html: `<ld-label size="m">Yada-yada</ld-label>`,
    })
    expect(root).toMatchSnapshot()
  })

  it('clicks on associated input element, when label is clicked', async () => {
    const { root } = await newSpecPage({
      components: [LdLabel],
      html: `<ld-label size="m">
        <span>Label text</span>
        <input type="text" />
      </ld-label>`,
    })
    const input = root.querySelector('input')

    input.click = jest.fn()
    input.focus = jest.fn()
    // A click on the label actually triggers a click on the slot in the shadow DOM.
    root.shadowRoot.querySelector('slot').click()
    root.querySelector('span').click()

    expect(input.click).toHaveBeenCalledTimes(1)
  })

  it('focuses associated input element, when label is clicked', async () => {
    const { root } = await newSpecPage({
      components: [LdLabel],
      html: `<ld-label size="m">
        <span>Label text</span>
        <input type="text" />
      </ld-label>`,
    })
    const input = root.querySelector('input')

    input.click = jest.fn()
    input.focus = jest.fn()
    // A click on the label actually triggers a click on the slot in the shadow DOM.
    root.shadowRoot.querySelector('slot').click()
    root.querySelector('span').click()

    expect(input.focus).toHaveBeenCalledTimes(1)
  })

  it('focuses associated input web component, when label is clicked', async () => {
    const { root } = await newSpecPage({
      components: [LdInput, LdLabel],
      html: `<ld-label size="m">
        <span>Label text</span>
        <ld-input />
      </ld-label>`,
    })
    const input = root.querySelector('ld-input')

    input.click = jest.fn()
    input.focus = jest.fn()
    Object.defineProperty(input, 'focusInner', { value: jest.fn() })
    // A click on the label actually triggers a click on the slot in the shadow DOM.
    root.shadowRoot.querySelector('slot').click()
    root.querySelector('span').click()

    expect(input.focus).not.toHaveBeenCalled()
    expect(input.focusInner).toHaveBeenCalledTimes(1)
  })
})
