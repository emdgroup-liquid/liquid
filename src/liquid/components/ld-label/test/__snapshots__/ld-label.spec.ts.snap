// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-label renders with position left 1`] = `
<ld-label position="left">
  <mock:shadow-root>
    <label class="ld-label ld-label--left">
      <slot></slot>
    </label>
  </mock:shadow-root>
  Yada-yada
</ld-label>
`;

exports[`ld-label renders with position right 1`] = `
<ld-label position="right">
  <mock:shadow-root>
    <label class="ld-label ld-label--right">
      <slot></slot>
    </label>
  </mock:shadow-root>
  Yada-yada
</ld-label>
`;

exports[`ld-label renders with size "m" 1`] = `
<ld-label size="m">
  <mock:shadow-root>
    <label class="ld-label ld-label--m">
      <slot></slot>
    </label>
  </mock:shadow-root>
  Yada-yada
</ld-label>
`;

exports[`ld-label renders with slot 1`] = `
<ld-label>
  <mock:shadow-root>
    <label class="ld-label">
      <slot></slot>
    </label>
  </mock:shadow-root>
  Yada-yada
</ld-label>
`;
